### PR TITLE
Replace custom errors with Play errors

### PR DESCRIPTION
**Pull request by mario-galic**
Removed `ApiErrors` factory as Play Framework already has these errors (`NotFound`, `Unautorized`, `BadRequest`, `InternalServerError`), so no need to reinvent the wheel.

Replaced current `ApiResponse` custom abstraction which consists of 90 lines of code, with `type ApiResponse[T] = Future[Either[ApiError,T]]`. Now just wrap the result in `EitherT` monad and you can fold on left and right results.

These two changes resulted in about 200 LOC less.

### CODE REVIEW

File: app/controllers/UsersController.scala [hunk 1]
```diff
@@ -30,19 +29,20 @@ class UsersController @Inject() (
...
-      else if (limit.getOrElse(0) < 0) {
-        ApiResponse.Left(ApiErrors.badRequest("limit must be a positive integer"))
-      }
-      else if (query.length < MinimumQueryLength) {
-        ApiResponse.Left(ApiErrors.badRequest(s"query must be a minimum of $MinimumQueryLength characters"))
-      }
-      else {
-        userService.search(query, limit, offset)
-      }
+    if (offset.getOrElse(0) < 0) {
```
Comment (jfsoul) [inline app/controllers/UsersController.scala]: I know this hasn't changed but I prefer e.g. `offset.forall(_ < 0)` for these checks

File: app/repositories/ReservedUserNameWriteRepository.scala [hunk 2]
```diff
@@ -23,22 +23,23 @@ class ReservedUserNameWriteRepository @Inject() (environment: play.api.Environme
       findOne(MongoDBObject("username" -> username))
     } match {
       case Success(Some(r)) => Right(r)
-      case Success(None) => Left(ApiErrors.notFound)
-      case Failure(t) =>
-        logger.error(s"Could find reserved username: $username", t)
-        Left(ApiErrors.internalError(t.getMessage))
+      case Success(None) => Left(ApiError("Username not found"))
+      case Failure(error) =>
+        val title = s"Could find reserved username $username"
```
Comment (jfsoul) [inline app/repositories/ReservedUserNameWriteRepository.scala]: Could not?

File: app/repositories/ReservedUserNameWriteRepository.scala [hunk 3]
```diff
@@ -23,22 +23,23 @@ class ReservedUserNameWriteRepository @Inject() (environment: play.api.Environme
...
         case Right(r) => Try {
           remove(r)
         } match {
-          case Success(success) =>
-            loadReservedUsernames
+          case Success(success) => loadReservedUsernames
           case Failure(t) =>
-            logger.error(s"Could remove reserved username: $username", t)
-            Left(ApiErrors.internalError(t.getMessage))
+            val title = s"Could remove reserved username $username"
```
Comment (jfsoul) [inline app/repositories/ReservedUserNameWriteRepository.scala]: Could not?

File: app/services/UserService.scala [hunk 4]
```diff
@@ -205,33 +201,30 @@ class UserService @Inject() (usersReadRepository: UsersReadRepository,
...
       case Left(r) => Left(r)
     }
-    ApiResponse.Async(Future.successful(result))
   }
 
-  def validateEmail(user: User): ApiResponse[Boolean] = {
-    val result = doValidateEmail(user, emailValidated = true)
-    ApiResponse.Async(Future.successful(result))
+  def validateEmail(user: User): ApiResponse[Boolean] = Future {
+    doValidateEmail(user, emailValidated = true)
```
Comment (jfsoul) [inline app/services/UserService.scala]: Looks simple enough to inline

File: app/services/UserService.scala [hunk 5]
```diff
@@ -250,15 +243,13 @@ class UserService @Inject() (usersReadRepository: UsersReadRepository,
...
       case Left(r) => Future.successful(Left(r))
     }
-    ApiResponse.Async(result)
   }
 
-  def unsubscribeFromMarketingEmails(email: String): ApiResponse[User] = {
-    val result = usersWriteRepository.unsubscribeFromMarketingEmails(email)
-    ApiResponse.Async(Future.successful(result))
+  def unsubscribeFromMarketingEmails(email: String): ApiResponse[User] = Future {
+    usersWriteRepository.unsubscribeFromMarketingEmails(email)
```
Comment (jfsoul) [inline app/services/UserService.scala]: As above

**Comment by mario-galic** [comment]
@jfsoul I fixed the typo in error messages. I'll open a separate PR for the remaining review changes.

File: app/services/UserService.scala (hunk 5 above)
Comment (mario-galic) [inline app/services/UserService.scala]: The client of `unsubscribeFromMarketingEmails()` does not currently have access to `usersWriteRepository`, so I would need to inject it first.

File: app/services/UserService.scala (hunk 4 above)
Comment (mario-galic) [inline app/services/UserService.scala]: Done.

**Comment by jfsoul** [comment]
Great stuff by the way! 💯

**Pull request merged**